### PR TITLE
Authorize topics pub/sub if a parent level has permissions

### PR DESCRIPTION
**Pull request by akalankapagoda**
This pull fixes the following issues.

https://wso2.org/jira/browse/MB-1478
https://wso2.org/jira/browse/MB-1481
https://wso2.org/jira/browse/MB-1505

### CODE REVIEW

File: components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java [hunk 1]
```diff
@@ -554,6 +552,36 @@ public static Result handlePublishToExchange(String username, UserRealm userReal
...
+     *
+     * @return True if permission is given to any of the parent levels
+     *
+     * @throws UserStoreException
+     */
+    private static boolean isAuthorizedFromHierarchy(UserRealm userRealm, String username, String permission) throws
+            UserStoreException {
+        boolean authorized = false;
+
+        if (userRealm.getAuthorizationManager().isUserAuthorized(username, permission, TreeNode.Permission.PUBLISH
```
Comment (indikasampath2000) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java]: Shall we rename permission parameter to resourceId or resourcePath ? Just a suggestion..

File: components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java [hunk 2]
```diff
@@ -554,6 +552,36 @@ public static Result handlePublishToExchange(String username, UserRealm userReal
...
+                .toString().toLowerCase())) { // Check if the current level has permission
+            authorized = true;
+        } else { // Check if the parent level has permission
+            int lastDelimiterIndex = permission.lastIndexOf("/");
+
+            if (lastDelimiterIndex > 0) { // Check parent only if a parent level is available
+                String parent = permission.substring(0, lastDelimiterIndex);
+                authorized = isAuthorizedFromHierarchy(userRealm, username, parent);
```
Comment (indikasampath2000) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java]: recursive call ? @ramith @a5anka  any comment ?

Comment (ramith) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java]: well, I don't see a problem. is there a possibility of going for a infinite recursion?

File: components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java (hunk 1 above)
Comment (a5anka) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java]: :+1: I too agree. permission is little bit confusing

File: components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java (hunk 2 above)
Comment (a5anka) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java]: I also don't see a problem. An alternative is to use a for loop.

Comment (akalankapagoda) [inline components/andes/org.wso2.carbon.andes.authorization/src/main/java/org/wso2/carbon/andes/authorization/andes/AndesAuthorizationHandler.java]: anyway we won't be using this method. Just found out that there is a similar method

**Comment by indikasampath2000** [comment]
:+1:

**Pull request merged**